hibernate-core/src/main/java/org/hibernate/persister/walking/spi/MetamodelGraphWalker.java

```diff
@@ -326,4 +326,16 @@ protected void addAssociationKey(AssociationKey associationKey) {
 	protected boolean isDuplicateAssociationKey(AssociationKey associationKey) {
 		return visitedAssociationKeys.contains( associationKey ) || strategy.isDuplicateAssociationKey( associationKey );
 	}
+
+	protected boolean isDuplicateAttributeDefinition(AssociationAttributeDefinition attributeDefinition) {
+		if ( isDuplicateAssociationKey( attributeDefinition.getAssociationKey() ) ) {
+			if ( attributeDefinition.getAssociationNature() == AssociationAttributeDefinition.AssociationNature.ENTITY ) {
+				return strategy.isDuplicateAssociatedEntity( attributeDefinition );
+			}
+			else {
+				return true;
+			}
+		}
+		return false;
```

Comment:
I think we can make the above logic a little bit more compact as
```
return !isDuplicateAssociationKey( attributeDefinition.getAssociationKey() ||
           attributeDefinition.getAssociationNature() != AssociationAttributeDefinition.AssociationNature.ENTITY ||
           strategy.isDuplicateAssociatedEntity( attributeDefinition );
```
Or a little bit less aggressive as
```
if ( isDuplicateAssociationKey( attributeDefinition.getAssociationKey() ) ) {
    return attributeDefinition.getAssociationNature() != AssociationAttributeDefinition.AssociationNature.ENTITY || strategy.isDuplicateAssociatedEntity( attributeDefinition );
}
else {
    return false;
}
```
Subjective and minor. Feel free to ignore this comment.

Reply:
@NathanQingyangXu I will add some comments here as well, and try to simplify the logic.